### PR TITLE
[Dist] Reduce peak memory in DistDGL

### PR DESCRIPTION
**Pull request by Rhett-Ying**
## Description

all the action items listed in https://github.com/dmlc/dgl/issues/4510 have been handled.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/distributed/dist_graph.py [hunk 1]
```diff
@@ -381,6 +377,10 @@ def __init__(self, server_id, ip_config, num_servers,
                 self.init_data(name=str(data_name), policy_str=data_name.policy_str,
                                data_tensor=node_feats[name])
                 self.orig_data.add(str(data_name))
+            del node_feats
+            gc.collect()
```
Comment (jermainewang) [inline python/dgl/distributed/dist_graph.py:381]: Look like L370-381 is to get the feature names but it loads the entire feature data instead. Is it possible to read them from some metadata files?

File: python/dgl/distributed/partition.py [hunk 2]
```diff
@@ -96,27 +114,27 @@ def load_partition(part_config, part_id, load_feats=True):
...
-        eids = F.boolean_mask(graph.edata[EID], _get_inner_edge_mask(graph, etype_id))
-        partids1 = gpb.eid2partid(eids)
-        _, per_type_eids = gpb.map_to_per_etype(eids)
-        partids2 = gpb.eid2partid(per_type_eids, etype)
-        assert np.all(F.asnumpy(partids1 == part_id)), 'load a wrong partition'
-        assert np.all(F.asnumpy(partids2 == part_id)), 'load a wrong partition'
-        etypes_list.append(etype)
+    ntypes_list = list(ntypes.keys())
+    etypes_list = list(etypes.keys())
+    if 'DGL_DIST_DEBUG' in os.environ:
```
Comment (jermainewang) [inline python/dgl/distributed/partition.py:119]: We really need a docpage about all the environment variables used in DistDGL. A temporary solution is to add a page to the distributed training user guide.

File: python/dgl/distributed/partition.py [hunk 3]
```diff
@@ -96,27 +114,27 @@ def load_partition(part_config, part_id, load_feats=True):
...
+        for ntype in ntypes:
+            ntype_id = ntypes[ntype]
+            # graph.ndata[NID] are global homogeneous node IDs.
+            nids = F.boolean_mask(graph.ndata[NID], _get_inner_node_mask(graph, ntype_id))
+            partids1 = gpb.nid2partid(nids)
+            _, per_type_nids = gpb.map_to_per_ntype(nids)
+            partids2 = gpb.nid2partid(per_type_nids, ntype)
+            assert np.all(F.asnumpy(partids1 == part_id)), 'load a wrong partition'
+            assert np.all(F.asnumpy(partids2 == part_id)), 'load a wrong partition'
```
Comment (jermainewang) [inline python/dgl/distributed/partition.py:128]: Please change them to more informative error message.

File: python/dgl/distributed/partition.py [hunk 4]
```diff
@@ -96,27 +114,27 @@ def load_partition(part_config, part_id, load_feats=True):
...
+        for etype in etypes:
+            etype_id = etypes[etype]
+            # graph.edata[EID] are global homogeneous edge IDs.
+            eids = F.boolean_mask(graph.edata[EID], _get_inner_edge_mask(graph, etype_id))
+            partids1 = gpb.eid2partid(eids)
+            _, per_type_eids = gpb.map_to_per_etype(eids)
+            partids2 = gpb.eid2partid(per_type_eids, etype)
+            assert np.all(F.asnumpy(partids1 == part_id)), 'load a wrong partition'
+            assert np.all(F.asnumpy(partids2 == part_id)), 'load a wrong partition'
```
Comment (jermainewang) [inline python/dgl/distributed/partition.py:137]: Same here.

File: python/dgl/distributed/partition.py [hunk 5]
```diff
@@ -13,6 +13,24 @@
 from ..partition import metis_partition_assignment, partition_graph_with_halo, get_peak_mem
 from .graph_partition_book import BasicPartitionBook, RangePartitionBook
 
+FIELD_DICT = {'inner_node': F.uint8,    # A flag indicates whether the node is inside a partition.
```
Comment (jermainewang) [inline python/dgl/distributed/partition.py:16]: Let's have a more readable name. How about `RESERVED_FIELD_DTYPE` ?

File: python/dgl/distributed/partition.py [hunk 6]
```diff
@@ -873,10 +901,10 @@ def get_homogeneous(g, balance_ntypes):
...
                     else:
                         edge_feats[etype + '/' + name] = g.edges[etype].data[name]
-        # Some adjustment for heterogeneous graphs.
-        if not g.is_homogeneous:
-            part.ndata['orig_id'] = F.gather_row(sim_g.ndata[NID], part.ndata['orig_id'])
-            part.edata['orig_id'] = F.gather_row(sim_g.edata[EID], part.edata['orig_id'])
+        # delete `orig_id` from ndata/edata
+        if reshuffle:
+            del part.ndata['orig_id']
+            del part.edata['orig_id']
```
Comment (jermainewang) [inline python/dgl/distributed/partition.py:907]: Instead of just patching it at the end, could you take the chance to refactor the code too? For example, could we not generate/save this `'orig_id'` at the beginning?

File: tools/distpartitioning/utils.py [hunk 7]
```diff
@@ -390,7 +390,7 @@ def write_graph_dgl(graph_file, graph_obj):
     graph_file : string
         File name in which graph object is serialized
     """
-    dgl.save_graphs(graph_file, [graph_obj])
+    dgl.distributed.partition._save_graphs(graph_file, [graph_obj])
```
Comment (jermainewang) [inline tools/distpartitioning/utils.py:393]: Is this change needed?

**Review by jermainewang — CHANGES_REQUESTED** [review]

File: python/dgl/distributed/dist_graph.py (hunk 1 above)
Comment (Rhett-Ying) [reply python/dgl/distributed/dist_graph.py:381]: no. feature data is really used in `self.init_data()`. we really need to load the feature data.

File: python/dgl/distributed/partition.py (hunk 2 above)
Comment (Rhett-Ying) [reply python/dgl/distributed/partition.py:119]: how about achieving this in another dedicated PR?

File: tools/distpartitioning/utils.py (hunk 7 above)
Comment (Rhett-Ying) [reply tools/distpartitioning/utils.py:393]: just to make sure the saved graph has expected dtype.

File: python/dgl/distributed/partition.py (hunk 6 above)
Comment (Rhett-Ying) [reply python/dgl/distributed/partition.py:907]: orig_nids/eids are required during update `part_g.ndata[NTYPE]/edata[ETYPE]`. so we have to generate. I think we could avoid attach such data into `part_g.n/edata`.

File: python/dgl/distributed/partition.py (hunk 5 above)
Comment (Rhett-Ying) [reply python/dgl/distributed/partition.py:16]: ok

File: python/dgl/distributed/partition.py (hunk 2 above)
Comment (jermainewang) [reply python/dgl/distributed/partition.py:119]: sure

File: python/dgl/distributed/partition.py (hunk 6 above)
Comment (Rhett-Ying) [reply python/dgl/distributed/partition.py:918]: seems `'orig_id'` is always attached if `reshuffle=True` as shown below.
https://github.com/dmlc/dgl/blob/303b150f8c6b6dbd6ea2b3ca724ded64ae14cb5f/python/dgl/partition.py#L140-L152

Comment (Rhett-Ying) [reply python/dgl/distributed/partition.py:918]: 'orig_id' is useful and user may rely on it directly. as a result, I think we'd better leave `orig_id` in `part_g.n/edata`.

Comment (jermainewang) [reply python/dgl/distributed/partition.py:918]: ok

**Review by jermainewang — APPROVED** [review]
LGTM

File: python/dgl/distributed/partition.py (hunk 6 above)
Comment (chwan1016) [reply python/dgl/distributed/partition.py:918]: @Rhett-Ying Can I revert the change of these lines? 'orig_id' is really useful for me to debug a code. With this key, I can verify whether the node embedding in a subgraph is identical to the result in the original graph.

Comment (Rhett-Ying) [reply python/dgl/distributed/partition.py:918]: @chwan1016 I'm afraid not. You need to change your test logic: fetch such info from the saved files.

Comment (chwan1016) [reply python/dgl/distributed/partition.py:918]: @Rhett-Ying Could you please be more specific? `orig_id` is deleted in this piece of code and it does not exist in any file. Thanks.

Comment (chwan1016) [reply python/dgl/distributed/partition.py:918]: Oh, great! I didn't notice the `return_mapping` param before. Thanks!

Comment (chwan1016) [reply python/dgl/distributed/partition.py:918]: @Rhett-Ying Can we replace `return_mapping` with another param like `store_mapping` (similar with `store_id` in `node_subgraph()`)? Storing the `orig_id` with other graph data together would make the file structure cleaner. Thanks!

Comment (Rhett-Ying) [reply python/dgl/distributed/partition.py:918]: It's intended to store `orig_id` separately from graph data as we don't always need such info together with graph data when loading a graph for distributed train even we saved such info.
